tools/hardhat-example/test/discrepancies.test.ts

```diff
@@ -174,34 +174,56 @@ describe('Demo discrepancies between hardhat and hedera local node', function()
     const hERC20UtilContract: HERC20Util = (await HERC20UtilFactory.deploy({ gasLimit: 5_000_000 })) as HERC20Util;
     await hERC20UtilContract.deployed();
 
+    // async function getTokenBalances(address: string): Promise<Array<BigNumber>> {
+    //   const balanceTokenA = erc20TokenA.balanceOf(address);
+    //   const balanceTokenB = erc20TokenB.balanceOf(address);
+    //
+    //   return Promise.all([
+    //     balanceTokenA,
+    //     balanceTokenB
+    //   ]);
+    // }
+    //
+    // async function getTokenBalance(tokens: string[], address: string) {
+    //   const promisesA: Promise<BigNumber>[] = [];
+    //   const promisesB: Promise<BigNumber>[] = [];
+    //   for (let token of tokens) {
+    //     promisesA.push(hERC20UtilContract.balanceOf(token, address));
+    //     promisesB.push(balanceOf(token, address));
+    //   }
+    //   const A = await Promise.all(promisesA);
+    //   const B = await Promise.all(promisesB);
+    //
+    //   for (let index in A) {
+    //     const balanceA = A[index];
+    //     const balanceB = B[index];
+    //
+    //     // for some reason the balance returned by the HERC20Util is not correct; but the balance returned via the @hashgraph/sdk is correct; hence why B and NOT A is returned
+    //     if (!balanceA.eq(balanceB)) {
+    //       console.log("balance Core !== balance HERC20Util");
+    //     }
+    //   }
+    //
+    //   return B;
+    // }
+
     async function getTokenBalances(address: string): Promise<Array<BigNumber>> {
```

Comment:
Hi @natanasow thanks for taking the time to investigate this issue.

As far as I can tell both of our implementations(for both `getTokenBalances` and `getTokenBalance` functions) should be functionally equivalent and should return the same data.

Since yours does not have the same issue, it would seem that it exposes a race condition on the hedera local node as yours first waits for the 1st balance to resolve before getting the 2nd token balance, and by the time the 2nd token balance is requested sufficient time has passed that the hedera local node will now return the latest and correct balance.

If you had to implement an ERC20 version of the `SimpleVault` contract and run an equivalent implementation of my test case against the hardhat local node, there'd be no issue.(I can confirm as I've done this with another contract.)